### PR TITLE
[IMP] mail: improve the way message are sent

### PR DESCRIPTION
**Pull request by phenix-factory**
The code that is used to sent message is splitted between composer and
composer_view model.
This PR centralize the sending process inside a message_composition model.

The body of a message is voluntarily NOT a compute to avoid possible performances
issues.

Part of task-2390637

### CODE REVIEW

**Comment by phenix-factory** [comment]
@seb-odoo @alexkuhn This should be good for review now :)

**Comment by seb-odoo** [comment]
You'll need to rebase and adapt to new models naming.

**Comment by phenix-factory** [comment]
@seb-odoo @alexkuhn This PR is rebased and green :tada:

File: addons/mail/static/src/models/message_composition/message_composition.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+         */
+        rawBody: attr({
+            compute: '_computeRawBody',
+        }),
+        /**
+         * States the mentionned channel in the rawBody.
+         */
+        mentionedChannels: many2many('Thread'),
+        /**
+         * States the mentionned partners in the rawBody.
```
Comment (alexkuhn) [inline addons/mail/static/src/models/message_composition/message_composition.js:199]: Just "body" (instead of "rawBody") is good.
"rawBody" gives the impression these fields are more technical that they really are.

**Review by alexkuhn — COMMENTED** [review]
At a glance, this is essentially a move of concern from `Composer` to dedicated model `MessageComposition`.

It's good to make lots of small models. A bit annoying we have to deal with OOP composition for the time being, but hopefully this will all be solved with traits in the future.

File: addons/mail/static/src/models/message_composition/message_composition.js (hunk 1 above)
Comment (phenix-factory) [reply addons/mail/static/src/models/message_composition/message_composition.js]: Done !

**Comment by phenix-factory** [comment]
@seb-odoo @alexkuhn This PR is getting old. Does anything still need to be done ?

File: addons/mail/static/src/models/composer_view/composer_view.js [hunk 2]
```diff
@@ -132,23 +131,23 @@ registerModel({
...
+                    this.composer.messageComposition.update({ mentionedPartners: link(this.activeSuggestedRecord) });
                     break;
             }
+            const updateData = {
+                isLastStateChangeProgrammatic: true,
+                textInputContent: textLeft + recordReplacement + ' ' + textRight,
+                textInputCursorEnd: textLeft.length + recordReplacement.length + 1,
+                textInputCursorStart: textLeft.length + recordReplacement.length + 1,
+            };
             this.composer.update(updateData);
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view/composer_view.js:151]: Could just `composer.update({ ... })`. The variable `updateData` is no more necessary.

File: addons/mail/static/src/models/message_composition/message_composition.js [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        thread: one2one('Thread', {
+            compute: '_computeThread',
+        }),
+        isLog: attr({
+            compute: '_computeIsLog',
+        }),
+        recipients: many2many('Partner', {
+            compute: '_computeRecipients',
+        }),
+    }
```
Comment (alexkuhn) [inline addons/mail/static/src/models/message_composition/message_composition.js:211]: alphabetical order (also compute methods)

File: addons/mail/static/src/models/message_composition/message_composition.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
+/** @odoo-module **/
+
+import { emojis } from '@mail/js/emojis';
+import { registerModel } from '@mail/model/model_core';
+import { unlink, replace } from '@mail/model/model_field_command';
+import { attr, many2many, one2one } from '@mail/model/model_field';
+import { addLink, escapeAndCompactTextContent, parseAndTransform } from '@mail/js/utils';
+
+registerModel({
+    name: 'MessageComposition',
```
Comment (alexkuhn) [inline addons/mail/static/src/models/message_composition/message_composition.js:10]: We didn't take the habit to put docstring above new model register.
Could be worth start doing it. Same motivation as with some fields that are one-line comments: `camelCase` naming are not as easy to get as it seems, so a plain English text helps quite a lot to understand things quickly.

For instance, just telling that this model represent a message being composed helps already.

File: addons/mail/static/src/models/composer_view/composer_view.js [hunk 5]
```diff
@@ -247,22 +246,14 @@ registerModel({
...
-            // takes the responsibility to linkify the urls we end up with double
-            // linkification a bit everywhere. Ideally we want to keep the content
-            // as text internally and only make html enrichment at display time but
-            // the current design makes this quite hard to do.
-            body = this._generateMentionsLinks(body);
-            body = parseAndTransform(body, addLink);
-            body = this._generateEmojisOnHtml(body);
+
+            composer.messageComposition.generateBody();
+
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view/composer_view.js:251]: no newline before/after

File: addons/mail/static/src/components/file_uploader/file_uploader.js [hunk 6]
```diff
@@ -97,6 +97,7 @@ export class FileUploader extends Component {
                 mimetype: file.type,
                 name: file.name,
                 originThread: (!composer && thread) ? replace(thread) : undefined,
+                messageCompositions: composer && composer.messageComposition && link(composer.messageComposition),
```
Comment (alexkuhn) [inline addons/mail/static/src/components/file_uploader/file_uploader.js:100]: should delete `composer:` line 93 (above)

File: addons/mail/static/src/components/file_uploader/file_uploader.js [hunk 7]
```diff
@@ -151,6 +152,7 @@ export class FileUploader extends Component {
         }
         const attachment = this.messaging.models['Attachment'].insert({
             composer: composer && replace(composer),
+            messageCompositions: composer && composer.messageComposition && link(composer.messageComposition),
```
Comment (alexkuhn) [inline addons/mail/static/src/components/file_uploader/file_uploader.js:155]: same as before, should delete `composer:`

There should be a crash, but it doesn't happen. I think @oomsveta applied a fix recently, so if you rebase you should have proper crash.

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js [hunk 8]
```diff
@@ -215,6 +215,7 @@ export class ComposerTextInput extends Component {
         }
         this._textareaLastInputValue = this._textareaRef.el.value;
         this._updateHeight();
+        this.composerView.composer.messageComposition.removeSuggestions();
```
Comment (alexkuhn) [inline addons/mail/static/src/components/composer_text_input/composer_text_input.js:218]: Does it fix a bug?

Could be another PR in v14 or v15 if that's the case, and this line would simply be a move from `Composer/removeSuggestions` to `MessageComposition/removeSuggestions` (which makes sense in the context of this PR).

**Review by alexkuhn — CHANGES_REQUESTED** [review]
I really like this PR slightly reducing complexity of `Composer` by introducing `MessageComposition`.

File: addons/mail/static/src/components/composer_text_input/composer_text_input.js (hunk 8 above)
Comment (phenix-factory) [reply addons/mail/static/src/components/composer_text_input/composer_text_input.js:220]: It fix a bug introduced by the PR :)

Comment (phenix-factory) [reply addons/mail/static/src/components/composer_text_input/composer_text_input.js]: You are right, in fact, it's doesn't look useful at all. I don't remember exactly why I did it in the first place, but I guess it doesn't matter anymore :)

File: addons/mail/static/src/models/composer_view/composer_view.js [hunk 9]
```diff
@@ -515,12 +503,16 @@ registerModel({
...
+         * @returns {FieldCommand}
          */
         _computeAttachmentList() {
-            return (this.composer && this.composer.attachments.length > 0)
-                ? insertAndReplace()
-                : clear();
+            if (this.composer &&
+                this.composer.messageComposition &&
+                this.composer.messageComposition.attachments.length > 0
+               ) {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view/composer_view.js:512]:
```js
if (
    A &
    B &
    C
) {
}
```

File: addons/mail/static/src/models/composer_view/composer_view.js (hunk 5 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/composer_view/composer_view.js:324]: still todo: no newline before/after

**Comment by phenix-factory** [comment]
@alexkuhn @seb-odoo This PR is green now.

File: addons/mail/static/src/models/composer_view/composer_view.js [hunk 10]
```diff
@@ -509,12 +493,17 @@ registerModel({
...
          */
         _computeAttachmentList() {
-            return (this.composer && this.composer.attachments.length > 0)
-                ? insertAndReplace()
-                : clear();
+            if (
+                this.composer &&
+                this.composer.messageComposition &&
+                this.composer.messageComposition.attachments.length > 0
+               ) {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer_view/composer_view.js:503]: unindent once

File: addons/mail/static/src/models/message_composition/message_composition.js (hunk 4 above)
Comment (alexkuhn) [reply addons/mail/static/src/models/message_composition/message_composition.js:10]: Would be nice a small comment that briefly explain what this model is about.
Something like:

```
Model of a message being composed in a composer.

Note that a message composition is not a full-fledged message,
because it hasn't been posted yet.
Still, it looks somewhat like a message, which is intended because
when a message composition is posted, it becomes a message.
```

1st paragraph just in plain english what's this model, and 2nd paragraph to clarify nuance between message composition and message.

File: addons/mail/static/src/models/message_composition/message_composition.js [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+            compute: '_computeRecipients',
+        }),
+        /**
+         * States the mentionned channel in the body.
+         */
+        mentionedChannels: many2many('Thread'),
+        /**
+         * States the mentionned partners in the body.
+         */
+        mentionedPartners: many2many('Partner'),
```
Comment (alexkuhn) [inline addons/mail/static/src/models/message_composition/message_composition.js:175]: `mentionedX` fields before `rawBody`

File: addons/mail/static/src/models/message_composition/message_composition.js [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+         */
+        _computeRawBody() {
+            return this.composer.textInputContent;
+        },
+        /**
+         * @private
+         * @returns {FieldCommand}
+         */
+        _computeThread() {
+            return replace(this.composer.activeThread);
```
Comment (alexkuhn) [inline addons/mail/static/src/models/message_composition/message_composition.js:135]: I think it's possible that `composer.activeThread` is undefined (due to rudimentary internal code of models), and `replace(undefined)` makes a crash.

So best to guard this.

**Review by alexkuhn — APPROVED** [review]
LGTM

**Comment by alexkuhn** [comment]
@seb-odoo I'd like to merge this PR. Is it ok for you?

**Comment by seb-odoo** [comment]
If it's that close to merge I'll make a last check today then

**Review by seb-odoo — CHANGES_REQUESTED** [review]
I don't really see how message composition is different than composer. If done properly almost everything from composer will be moved to this new model, so it's just a rename.

Ignoring the rename for now, it looks like what need to be done in the queue PR is just to create one "composer" per new message, instead of reusing the same one. (the composer view will be re-used now that we have it, but composer is just data).

See remarks below for some pointers, but I'm not sure if it's worth fixing them, instead of just flat out renaming, and maybe moving anything that would be "problematic" for the other PR in composer view instead of composer.
